### PR TITLE
Support GroupPartitionsExec on GPU [databricks][fast-ut][reduced-it]

### PR DESCRIPTION
**Pull request by firestarman**
Part of #15602.

### Description

- Add a Spark 4.2 `GpuGroupPartitionsExec` wrapper that reuses Spark's grouping and coalescing metadata so non-sorted storage-partitioned joins remain columnar without changing Spark semantics.
- Capture grouping metadata before child conversion and keep sorted-merge plans on the original CPU subtree because row/columnar transitions can hide `KeyedPartitioning` and GPU k-way sorted merge is not yet supported.
- Consolidate Spark 4.2 unit coverage for grouped partitions, empty groups, child transitions, bounded plan summaries, and sorted-merge fallback so both supported and fallback paths are protected.
- Add a deterministic V2 integration source and cross-version test for coalescing, replication, empty padding, and CPU-scan transitions because Spark 4.2 does not yet have a compatible accelerated Iceberg path for this scenario.
- Validated `GroupPartitionsExecSuite` with Spark 4.2, `test_group_partitions_partial_clustering_distinct` with Spark 4.1.3 and 4.2, and `test_iceberg_spj_partial_clustering_distinct` with Spark 4.0.4.
- Validated integration-test packaging for build versions 330, 359, 404, 413, and 420, plus `mvn -s ~/.m2/settings_art.xml -f scala2.13/pom.xml -Dbuildver=420 -Dcuda.version=cuda13 -DskipTests -pl integration_tests -am validate`.

#### Performance

- Workload: Spark 4.2 partial-clustering SPJ plus `DISTINCT`, comparing `NVDA/main` commit `80bc76bfc` CPU fallback with the PR GPU wrapper; RAPIDS was enabled for both.

  ```sql
  SELECT DISTINCT l.id
  FROM group_partitions_left l
  JOIN group_partitions_right r ON l.id = r.id
  ```

- Setup: Scala 2.13, `local[4]`, NVIDIA RTX 5880 Ada, 8 GiB driver heap, AQE and broadcast joins disabled, 20 million scanned left rows, and 15 million join-output rows.
- Runs: Four separate Spark processes were run in this order: baseline, PR, baseline, PR. Each process ran two untimed warm-ups followed by five timed actions, giving 10 timed samples per variant.

| Variant | GroupPartitionsExec | Median action time | Speedup |
| --- | ---: | ---: | ---: |
| `NVDA/main` (`80bc76bfc`) | CPU | 1.495 s | 1.00x |
| PR (`2ab55439a`) | GPU | 0.658 s | 2.27x |

### Checklists

Documentation
- [ ] Updated for new or modified user-facing features or behaviors
- [x] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [x] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [ ] Not required

### CODE REVIEW

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

File: sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    GpuMetric.OP_TIME_NEW ->
+      createNanoTimingMetric(GpuMetric.MODERATE_LEVEL, GpuMetric.DESCRIPTION_OP_TIME_NEW))
+
+  override def output = child.output
+
+  override def outputPartitioning: Partitioning = groupInfo.outputPartitioning
+
+  override def outputOrdering: Seq[SortOrder] = groupInfo.outputOrdering
+
+  override def outputBatching: CoalesceGoal = GpuExec.outputBatching(child)
```
Comment (wjxiz1992) [inline sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala]: P1: This batching guarantee is not preserved when a group contains more than one parent partition. `CoalescedRDD` flat-maps the parent iterators, so a child that provides `RequireSingleBatch` per original task can produce multiple batches in the regrouped task. Because `optimizeCoalesce` removes `GpuCoalesceBatches` whenever `outputBatching` satisfies the requested goal, forwarding the child goal here can remove a coalesce that a downstream operator requires for correctness.

Please return no guarantee for multi-parent groups (or preserve only a provably safe `TargetSize` guarantee), and add a regression with a child advertising `RequireSingleBatch` and a group containing multiple parents.

File: integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/datasourcev2/GroupPartitionsDataSource.scala [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+/**
+ * A deterministic row-based V2 source for storage-partitioned join tests.
+ *
+ * It reports identity partitioning by `id` and exposes each entry below as a separate input
+ * partition. Keeping the source row-based also exercises the transition between a CPU scan and
+ * GpuGroupPartitionsExec.
+ */
+object GroupPartitionsDataSource {
+  val schema = StructType(Array(
```
Comment (res-life) [inline integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/datasourcev2/GroupPartitionsDataSource.scala]: Could we rename these companion-object fixtures to `SCHEMA`, `LEFT_PARTITIONS`, and `RIGHT_PARTITIONS`, and update their references? Immutable companion-object constants use `UPPER_SNAKE_CASE` elsewhere in the project.

File: sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import org.apache.spark.sql.catalyst.expressions.SortOrder
+import org.apache.spark.sql.catalyst.plans.physical.Partitioning
+import org.apache.spark.sql.catalyst.util.{truncatedString, InternalRowComparableWrapper}
+import org.apache.spark.sql.connector.catalog.functions.Reducer
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.datasources.v2.{GroupedPartitionCoalescer,
+  GroupPartitionsExec}
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+class GroupPartitionsExecMeta(
```
Comment (res-life) [inline sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala]: Could this be named `GpuGroupPartitionsExecMeta` to match the naming convention for metadata wrappers that create GPU replacement nodes? Please update the registration and test references at the same time.

File: sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    GpuGroupPartitionsExec(
+      childPlans.head.convertIfNeeded(),
+      groupInfo)
+  }
+}
+
+case class GpuGroupPartitionsExecInfo(
+    outputPartitioning: Partitioning,
+    outputOrdering: Seq[SortOrder],
+    groupedPartitions: Seq[(InternalRowComparableWrapper, Seq[Int])],
```
Comment (res-life) [inline sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala]: Can `GpuGroupPartitionsExecInfo` retain only the data used during GPU execution: `Seq[Seq[Int]]` for the partition groups and `Option[Int]` for the expected-key count? The execution path immediately maps `groupedPartitions` to `_._2`, while `expectedPartitionKeys` is used only for `.size`; keeping both complete keyed structures increases driver-plan memory and repeats the mapping allocation.

File: sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+      groupPartitions.joinKeyPositions,
+      groupPartitions.expectedPartitionKeys,
+      groupPartitions.reducers,
+      groupPartitions.distributePartitions)
+  }
+}
+
+case class GpuGroupPartitionsExec(
+    child: SparkPlan,
+    @transient groupInfo: GpuGroupPartitionsExecInfo)
```
Comment (res-life) [inline sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala:88]: Could we override `doCanonicalize` for the expressions captured inside `groupInfo`? `QueryPlan.mapExpressions` does not traverse an arbitrary nested case class, so `outputPartitioning` and `outputOrdering` can retain analysis-specific expression IDs while the child is canonicalized. Please canonicalize the child and normalize these captured expressions, with a regression test showing equivalent nodes with different expression IDs have equal canonical forms.

File: tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    assert(gpuPlan.collect { case exchange: Exchange => exchange }.isEmpty,
+      s"Storage-partitioned join unexpectedly contains an exchange:\n$gpuPlan")
+    assert(gpuGroups.exists(_.groupedPartitions.exists(_._2.size > 1)),
+      "Expected at least one key to coalesce multiple input partitions")
+    assert(gpuGroups.exists(_.groupedPartitions.exists(_._2.isEmpty)),
+      "Expected a missing key to produce an empty padded partition group")
+    val groupWithExpectedKeys = gpuGroups.find(_.expectedPartitionKeys.nonEmpty).getOrElse {
+      fail("Expected a GPU group with expected partition keys")
+    }
+    val summary = groupWithExpectedKeys.simpleString(maxFields = 1)
```
Comment (res-life) [inline tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala:77]: Could this coverage exercise every plan-summary field added by `planSummaryParts`? The current assertions check only `ExpectedPartitionKeys` and a total-length bound, so regressions that drop `JoinKeyPositions`, `Reducers`, or `DistributePartitions` would still pass. A focused metadata fixture with assertions for each label would cover those branches.

File: integration_tests/src/main/python/group_partitions_test.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+    assert len(scans) == 2, f"Expected two CPU batch scans, found {len(scans)}:\n{plan}"
+    assert len(joins) == 1, f"Expected one GPU SPJ join, found {len(joins)}:\n{plan}"
+
+    join_nodes = _collect_plan_nodes(joins[0])
+    join_exchanges = [
+        node for node in join_nodes
+        if node.getClass().getSimpleName() == "GpuShuffleExchangeExec"
+    ]
+    assert not join_exchanges, f"Expected shuffle-free SPJ inputs:\n{plan}"
```
Comment (res-life) [inline integration_tests/src/main/python/group_partitions_test.py:64]: This assertion checks only the join subtree. For Spark 4.2, could we also assert that `nodes_of_class("GpuShuffleExchangeExec")` is empty for the complete captured plan? That would detect an exchange inserted between the join and `DISTINCT`, where the test expects keyed partitioning to be reused.

**Review by res-life — COMMENTED** [review]
Leaving six inline comments covering style, performance, and test coverage. None is a result-correctness blocker.

File: sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala (hunk 3 above)
Comment (firestarman) [reply sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala]: Updated in b91619b8d. The metadata wrapper is now GpuGroupPartitionsExecMeta, including registration and test references.

File: sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala (hunk 1 above)
Comment (firestarman) [reply sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala]: Fixed in b91619b8d. Multi-parent groups now preserve only a TargetSize guarantee; other child batching guarantees are dropped. The regression covers RequireSingleBatch with multiple parents, the single-parent case, and TargetSize preservation.

File: integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/datasourcev2/GroupPartitionsDataSource.scala (hunk 2 above)
Comment (firestarman) [reply integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/datasourcev2/GroupPartitionsDataSource.scala]: Updated in b91619b8d. The companion-object fixtures are now SCHEMA, LEFT_PARTITIONS, and RIGHT_PARTITIONS, with all references updated.

File: sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala (hunk 4 above)
Comment (firestarman) [reply sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala]: Updated in b91619b8d. Group info now stores partition groups as Seq[Seq[Int]], the expected-key count as Option[Int], and reducer display names rather than complete keyed structures. The unused isGrouped copy was also removed.

File: sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala (hunk 5 above)
Comment (firestarman) [reply sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala:88]: Fixed in b91619b8d. doCanonicalize now canonicalizes the child and normalizes captured output partitioning and ordering expressions against the child output. A regression verifies equal canonical forms for equivalent nodes with different expression IDs.

File: tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala (hunk 6 above)
Comment (firestarman) [reply tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala:77]: Added in b91619b8d. A focused summary test now checks JoinKeyPositions, ExpectedPartitionKeys, Reducers, and DistributePartitions, while the existing integration assertion retains the bounded-length check.

File: integration_tests/src/main/python/group_partitions_test.py (hunk 7 above)
Comment (firestarman) [reply integration_tests/src/main/python/group_partitions_test.py:64]: Added in b91619b8d. The Spark 4.2 plan assertion now verifies that the complete captured plan contains no GpuShuffleExchangeExec, including the path between the join and DISTINCT.

**Review by res-life — APPROVED** [review]
LGTM

**Comment by firestarman** [comment]
build
